### ISSUE TITLE
How to save correct predictions of the val set in txt?

### ISSUE THREAD

**Issue by LorenzoSun-V** [THU-MIG/yolov10#78]
When I was doing val set inference, I saved the prediction results in txt, and found that all pictures had 300 detection results:
![image](https://github.com/THU-MIG/yolov10/assets/42413198/276cd88a-0dd8-46d2-8593-4ef1b05669a0)
![image](https://github.com/THU-MIG/yolov10/assets/42413198/1286ad0d-c304-4507-b6f9-40d8fdda0c05)
I used this command to do val set inference:
```
yolo val \
    model=${root}/${project}/${name}/weights/best.pt \
    data=${data} \
    imgsz=${imgsz} \
    batch=${batch}  \
    device=${device} \
    project=${project} \
    name=${name}/val \
    save_txt=true \
    save_conf=true \
    exist_ok=true \
```

What should I do if I want to save the correct predictions of the val set in txt?

**Comment by LorenzoSun-V** [comment]
In yolov8, using `save_txt` flag will only save the final results of the prediction(results after nms). However, the results after `v10postprocess` still have 300 predictions.
![image](https://github.com/THU-MIG/yolov10/assets/42413198/e882ffe5-17f1-48be-8466-02aef4082cb1)

**Comment by LorenzoSun-V** [comment]
In addition, you said `we adopt the top one selection`  in your paper:
![image](https://github.com/THU-MIG/yolov10/assets/42413198/89b234bf-6b10-4ee6-8bc6-f366c594fd53)

But the code seems to use the 300 predictions to calculate metrics with labels.
![image](https://github.com/THU-MIG/yolov10/assets/42413198/032a7c74-3337-4919-84bf-edfbc58af0b1)

**Comment by LorenzoSun-V** [comment]
How to understand this `dual label assignments`? It would be great if you could explain it in detail because I'm a little bit confused.  :star_struck:

**Comment by jameslahm** [comment]
Since there is no NMS in the postprocessing, we can directly output the `max_det` predictions.

> In yolov8, using `save_txt` flag will only save the final results of the prediction(results after nms). However, the results after `v10postprocess` still have 300 predictions. ![image](https://private-user-images.githubusercontent.com/42413198/334227954-e882ffe5-17f1-48be-8466-02aef4082cb1.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTY4NjAzODEsIm5iZiI6MTcxNjg2MDA4MSwicGF0aCI6Ii80MjQxMzE5OC8zMzQyMjc5NTQtZTg4MmZmZTUtMTdmMS00OGJlLTg0NjYtMDJhZWY0MDgyY2IxLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA1MjglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNTI4VDAxMzQ0MVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWMwYWYzNDE1ZTQ2MTM3NWM4ZDA2ODhmNDkwOTZmYWE5YTg3M2IwNDA3Y2IxNjNjNDA4NzIyNjFhMWQwNTJiNzcmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.EmBuCqPGso-wnNeQD7tF3OnG5EDVz9qJ11TTQzGZoAE)

Yes, during training, we adopt the top one selection to obtain one-to-one matching.
> In addition, you said we adopt the top one selection in your paper:

Thanks for your interest! In dual label assignments, we use one-to-one matching and one-to-many matching for the prediction at the same time. So, during training, there are two heads, namely one-to-many head and one-to-one head for optimization. During inference, we can leverage the one-to-one head to output predictions without NMS postprocessing.

> How to understand this dual label assignments? It would be great if you could explain it in detail because I'm a little bit confused. 🤩

**Comment by LorenzoSun-V** [comment]
`max_det` is equal to 300 which indicates there are 300 predictions after `v10postprocess`. And these 300 predictions are used to calculate mAP and other metrics. I suppose only the final predictions will be used in metrics calculation.
> Since there is no NMS in the postprocessing, we can directly output the `max_det` predictions.

**Comment by LorenzoSun-V** [comment]
I just wonder why don't use the same postprocess code of prediction in validation to calculate metrics. In other words, since there is no NMS process, why not directly use the results after threshold screening to calculate metrics.
![image](https://github.com/THU-MIG/yolov10/assets/42413198/df1c4b40-b46f-4df7-99aa-e2f9d6758eb9)

**Comment by LorenzoSun-V** [comment]
In my view, the code in red box should be added in the validator.
![image](https://github.com/THU-MIG/yolov10/assets/42413198/5ca7e3c8-2ada-4d5a-bf10-f93b7eea5978)

**Comment by jameslahm** [comment]
Thanks. Due to that the model does not need NMS, all output predictions can be directly added into the validator without the threshold.

**Comment by LorenzoSun-V** [comment]
In my case, the mAPs calculating by discarding the threshold are much lower than thosed calculating by adopting the threshold. Because a lot of useless predictions(predicitons with low scores, such as 1e-5) are employed to metrics calculation, and the neg samples in my validation set also have predictions. 

In COCO val set, the mAPs ​​are almost the same.
Discarding the threshold:
![image](https://github.com/THU-MIG/yolov10/assets/42413198/d78d4eca-a51f-41c2-8fc2-dca0b399cf77)
Adopting the threshold:
![image](https://github.com/THU-MIG/yolov10/assets/42413198/e0ec5707-6654-4707-9087-8bf55d76ec0c)

I just wonder why don't use the same confidence threshold setting like other models(such as 0.001) when validating. 
> Due to that the model does not need NMS, all output predictions can be directly added into the validator without the threshold.

**Comment by jameslahm** [comment]
Thanks. Do you calculate mAP in the same way as COCO?

**Comment by LorenzoSun-V** [comment]
Yes, I do. 
I used the same command as COCO to calculate mAP in my case.

**Comment by jameslahm** [comment]
Thanks. Could you please provide more details of how you use the same command as COCO to calculate mAP? Do you save the `predictions.json` and call `eval_json` to calculate the mAP? Thank you!

**Comment by LorenzoSun-V** [comment]
After enabling the `eval_json` using the "save_json" flag, the results are almost the same. Thanks.

**Comment by LorenzoSun-V** [comment]
Sorry, I made a mistake.

The details are as following.
### Case1. Adpoting the threshold
code:
```
class YOLOv10DetectionValidator(DetectionValidator):
    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)
        self.args.save_json |= self.is_coco

    def postprocess(self, preds):
        if isinstance(preds, dict):
            preds = preds["one2one"]

        if isinstance(preds, (list, tuple)):
            preds = preds[0]
        
        preds = preds.transpose(-1, -2)
        boxes, scores, labels = ops.v10postprocess(preds, self.args.max_det, self.nc)
        bboxes = ops.xywh2xyxy(boxes)
        # return torch.cat([bboxes, scores.unsqueeze(-1), labels.unsqueeze(-1)], dim=-1)
        preds = torch.cat([bboxes, scores.unsqueeze(-1), labels.unsqueeze(-1)], dim=-1)

        mask = preds[..., 4] > self.args.conf
        if self.args.classes is not None:
            mask = mask & (preds[..., 5:6] == torch.tensor(self.args.classes, device=preds.device).unsqueeze(0)).any(2)

        return [p[mask[idx]] for idx, p in enumerate(preds)]
```
command:
```
yolo val \
    model=${root}/${project}/${name}/weights/best.pt \
    data=${data} \
    imgsz=${imgsz} \
    batch=${batch}  \
    device=${device} \
    project=${project} \
    name=${name}/val_coco \
    conf=0.25 \
    save_json=true \
    plots=true
```
The results are:
```
Class     Images  Instances      Box(P          R      mAP50  mAP50-95)
     all       1161        990      0.915      0.915      0.951      0.712
```

### Case2. Discarding the threshold
code:
```
class YOLOv10DetectionValidator(DetectionValidator):
    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)
        self.args.save_json |= self.is_coco

    def postprocess(self, preds):
        if isinstance(preds, dict):
            preds = preds["one2one"]

        if isinstance(preds, (list, tuple)):
            preds = preds[0]
        
        preds = preds.transpose(-1, -2)
        boxes, scores, labels = ops.v10postprocess(preds, self.args.max_det, self.nc)
        bboxes = ops.xywh2xyxy(boxes)
        return torch.cat([bboxes, scores.unsqueeze(-1), labels.unsqueeze(-1)], dim=-1)
```
command:
```
yolo val \
    model=${root}/${project}/${name}/weights/best.pt \
    data=${data} \
    imgsz=${imgsz} \
    batch=${batch}  \
    device=${device} \
    project=${project} \
    name=${name}/val_coco \
    save_json=true \
    plots=true
```
The results are:
```
Class     Images  Instances      Box(P          R      mAP50  mAP50-95)
       all       1161        990      0.915      0.915      0.961      0.679
```

**Comment by jameslahm** [comment]
Thanks. Could you please manually set `self.is_coco = True` for the `validator` to keep the same way as COCO for calculating mAP?

**Comment by LorenzoSun-V** [comment]
I set `self.is_coco=True` manually and the difference is still existed. The results are as the same as before.

**Comment by LorenzoSun-V** [comment]
> In my case, the mAPs calculating by discarding the threshold are much lower than those calculating by adopting the threshold. Because a lot of useless predictions(predicitons with low scores, such as 1e-5) are employed to metrics calculation, and the neg samples in my validation set also have predictions.
> 
> In COCO val set, the mAPs ​​are almost the same. Discarding the threshold: ![image](https://private-user-images.githubusercontent.com/42413198/334606363-d78d4eca-a51f-41c2-8fc2-dca0b399cf77.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTY5NTI1MzMsIm5iZiI6MTcxNjk1MjIzMywicGF0aCI6Ii80MjQxMzE5OC8zMzQ2MDYzNjMtZDc4ZDRlY2EtYTUxZi00MWMyLThmYzItZGNhMGIzOTljZjc3LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA1MjklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNTI5VDAzMTAzM1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWJiYjNkODBkY2M0OTk0YzZhYTE0Njc3MzA3OTllNzlmY2JmOTE4MzQxNTE0YmMyMWI5NDg4OGFhMmViMzJlNmUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.wfcdyu-RnaDwVCi5eUZ9ZQhOcRrLKMe9MwjWzOxnPqs) Adopting the threshold: ![image](https://private-user-images.githubusercontent.com/42413198/334606389-e0ec5707-6654-4707-9087-8bf55d76ec0c.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTY5NTI1MzMsIm5iZiI6MTcxNjk1MjIzMywicGF0aCI6Ii80MjQxMzE5OC8zMzQ2MDYzODktZTBlYzU3MDctNjY1NC00NzA3LTkwODctOGJmNTVkNzZlYzBjLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA1MjklMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwNTI5VDAzMTAzM1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWJjMTUyMTViY2U1NDRlMjg2MGJmNDk4MDdlYWI1MGU4MDkwYmNiZmE4M2RiYzlhZmYxMzdjMTk5MzhiOWQ5ODgmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.MmGQuSrZGd4dNXv-pC5jzSQN6cKEWRJta3WLGMAF9Pg)
> 
> I just wonder why don't use the same confidence threshold setting like other models(such as 0.001) when validating.
> 
> > Due to that the model does not need NMS, all output predictions can be directly added into the validator without the threshold.

In this reply, I didn't use `conf` flag. `conf` is set as `0.001` as default.
After I use `conf=0.25` in COCO validation, the results are quite different:
![image](https://github.com/THU-MIG/yolov10/assets/42413198/fe405683-6194-4218-bc5e-ecf7212f7c84)

**Comment by LorenzoSun-V** [comment]
In my opinion, my task is much easier than COCO. So too much predictions are detrimental to the results. However, there are many hard and small samples in COCO so that using all output predictions to validation is better.

**Comment by jameslahm** [comment]
Thanks for the detailed explanation. Would you mind sharing your validation set and the checkpoint with us? We are trying to investigate this and confirm if there is an issue in the codebase. Thank you!

**Comment by LorenzoSun-V** [comment]
Of course! Would you mind giving your e-mail to me? Because the dataset is private.

**Comment by jameslahm** [comment]
Thank you! Our email is jameslahm17@gmail.com.

**Comment by leonnil** [comment]
Thanks for your assistance! We have successfully reproduced the issue. We believe the discrepancy can be attributed to the differences in mAP calculation methods between `YOLO` and `COCO`.

The `COCO` evaluation calculates mAP using the precision points where recall changes, while `YOLO` interpolates 101 points and performs integration by default. When evaluating with `COCO`, using the `conf` parameter may filter out true positive predictions, causing the precision to drop to 0, which negatively impacts the mAP on the COCO validation set. Conversely, `YOLO` uses interpolation to replace the precision of missing true positives, which might improve the mAP in some cases.

Therefore, we prefer not to set `conf` as a threshold to filter out predictions in order to maintain COCO mAP performance. We hope this explanation is helpful to you!

**Comment by LorenzoSun-V** [comment]
Yes, it does make sense. Thanks for your patient explanation.
